pkg/solana/chainwriter/lookups.go

```diff
@@ -0,0 +1,361 @@
+	if err != nil {
+		return nil, fmt.Errorf("error getting seeds for PDALookups: %w", err)
+	}
+
+	pdas, err := generatePDAs(publicKeys, seeds, pda)
+	if err != nil {
+		return nil, fmt.Errorf("error generating PDAs: %w", err)
+	}
+
+	if pda.InternalField.Location == "" {
+		return pdas, nil
+	}
+
+	// If a decoded location is specified, fetch the data at that location
+	var result []*solana.AccountMeta
+	for _, accountMeta := range pdas {
+		accountInfo, err := reader.GetAccountInfoWithOpts(ctx, accountMeta.PublicKey, &rpc.GetAccountInfoOpts{
+			Encoding:   "base64",
+			Commitment: rpc.CommitmentFinalized,
+		})
+
+		if err != nil || accountInfo == nil || accountInfo.Value == nil {
+			return nil, fmt.Errorf("error fetching account info for PDA account: %s, error: %w", accountMeta.PublicKey.String(), err)
+		}
+
+		decoded, err := decodeBorshIntoType(accountInfo.GetBinary(), pda.InternalField.Type)
+		if err != nil {
+			return nil, fmt.Errorf("error decoding Borsh data dynamically: %w", err)
+		}
+
+		value, err := GetValuesAtLocation(decoded, pda.InternalField.Location)
+		if err != nil {
+			return nil, fmt.Errorf("error getting value at location: %w", err)
+		}
+		if len(value) > 1 {
+			return nil, fmt.Errorf("multiple values found at location: %s", pda.InternalField.Location)
+		}
+
+		result = append(result, &solana.AccountMeta{
+			PublicKey:  solana.PublicKeyFromBytes(value[0]),
+			IsSigner:   accountMeta.IsSigner,
+			IsWritable: accountMeta.IsWritable,
+		})
+	}
+	return result, nil
+}
+
+func decodeBorshIntoType(data []byte, typ reflect.Type) (interface{}, error) {
+	// Ensure the type is a struct
+	if typ.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("provided type is not a struct: %s", typ.Kind())
+	}
+
+	// Create a new instance of the type
+	instance := reflect.New(typ).Interface()
+
+	// Decode using Borsh
+	err := ag_binary.NewBorshDecoder(data).Decode(instance)
+	if err != nil {
+		return nil, fmt.Errorf("error decoding Borsh data: %w", err)
+	}
+
+	// Return the underlying value (not a pointer)
+	return reflect.ValueOf(instance).Elem().Interface(), nil
+}
+
+// getSeedBytes extracts the seeds for the PDALookups.
+// It handles both AddressSeeds (which are public keys) and ValueSeeds (which are byte arrays from input args).
+func getSeedBytes(ctx context.Context, lookup PDALookups, args any, derivedTableMap map[string]map[string][]*solana.AccountMeta, reader client.Reader) ([][]byte, error) {
+	var seedBytes [][]byte
+	maxSeedLength := 32
+
+	for _, seed := range lookup.Seeds {
+		if lookupSeed, ok := seed.(AccountLookup); ok {
+			// Get value from a location (This doens't have to be an address, it can be any value)
+			bytes, err := GetValuesAtLocation(args, lookupSeed.Location)
+			if err != nil {
+				return nil, fmt.Errorf("error getting address seed: %w", err)
+			}
+			// validate seed length
+			for _, b := range bytes {
+				if len(b) > maxSeedLength {
```

Comment:
nit: noticed this while looking into PDAs for ChainReader but think the solana go package has public values we can use for this instead
```suggestion
				if len(b) > solana.MaxSeedLength {
```

Reply:
ah good find! fixed